src/dataset.rs

```diff
@@ -508,6 +511,84 @@ impl Dataset {
         }
         Ok(Transaction::new(self))
     }
+
+    /// Execute a SQL query against the Dataset. It is equivalent to calling
+    /// [`GDALDatasetExecuteSQL`](https://gdal.org/api/raster_c_api.html#_CPPv421GDALDatasetExecuteSQL12GDALDatasetHPKc12OGRGeometryHPKc).
+    /// Returns a `ResultSet`, which can be treated just as any other [`Layer`], or `None` if the
+    /// query does not have a `ResultSet`, such as `ALTER TABLE`, etc. Note that this is different
+    /// than a query which has an *empty* `ResultSet`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use gdal::Dataset;
+    /// # use std::path::Path;
+    /// #
+    /// let ds = Dataset::open(Path::new("fixtures/roads.geojson")).unwrap();
+    /// let query = "SELECT kind, is_bridge, highway FROM roads WHERE highway = 'pedestrian'";
+    /// let result_set = ds.execute_sql(query, None, None).unwrap().unwrap();
+    ///
+    /// assert_eq!(10, result_set.feature_count());
+    ///
+    /// for feature in result_set.features() {
+    ///     let highway = feature
+    ///         .field("highway")
+    ///         .unwrap()
+    ///         .unwrap()
+    ///         .into_string()
+    ///         .unwrap();
+    ///
+    ///     assert_eq!("pedestrian", highway);
+    /// }
+    /// ```
+    pub fn execute_sql<S: AsRef<str>>(
+        &self,
+        query: S,
+        spatial_filter: Option<&Geometry>,
+        dialect: Option<S>,
+    ) -> Result<Option<ResultSet>> {
+        let query = CString::new(query.as_ref())?;
+        let mut filter_geom: OGRGeometryH = std::ptr::null_mut();
+
+        let dialect_c_str = match dialect {
+            None => None,
+            Some(d) => Some(CString::new(d.as_ref())?),
+        };
+
+        let dialect_ptr = match dialect_c_str {
+            None => std::ptr::null(),
+            Some(d) => d.as_ptr(),
```

Comment:
```suggestion
            Some(ref d) => d.as_ptr(),
```

Damn, this is a subtle one.

Reply:
Holy crap, nice find! I actually don't understand this. What's going on here? Why does taking the `CString` by reference produce a different pointer than by value?

(Thanks for the review by the way! I'll address the comments when I have some more time.)

Reply:
Move along, nothing to see here...

---

The problem is that the `Some` branch of the `match` consumes `d` and thus `dialect_c_str` (if you duplicate the `match`, you'll get an error that `d` or `dialect_c_str` was moved out of). Then, `dialect_ptr` is a dangling pointer.

Reply:
Ahhhh, that is subtle! The dangers of unsafe code. Nice find.

Reply:
Yeah, apparently there's a lint for it, but only for simpler cases: https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#temporary-cstring-as-ptr